pkg/resources/util.go

```diff
@@ -0,0 +1,160 @@
+package resources
+
+import (
+	"context"
+
+	"github.com/integr8ly/cloud-resource-operator/pkg/apis/integreatly/v1alpha1"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// define a function that allows us to perform modification logic on the
+// custom resource (e.g. setting owner refs) before creating or updating it
+type modifyResourceFunc func(cr metav1.Object) error
+
+// ReconcileBlobStorage creates or updates a blob storage custom resource
+func ReconcileBlobStorage(ctx context.Context, client client.Client, resourceType, tier, name, ns, secretName, secretNs string, modifyFunc modifyResourceFunc) (*v1alpha1.BlobStorage, error) {
+	bs := &v1alpha1.BlobStorage{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: ns,
+		},
+	}
+
+	// execute logic to modify the resource before creation
+	// e.g. add owner refs
+	if modifyFunc != nil {
+		err := modifyFunc(bs)
+		if err != nil {
+			return nil, err
```

Comment:
can we wrap these errors to provide context?

Reply:
sure